default/python/AI/SaveGameManager/_decoder.py

```diff
@@ -0,0 +1,182 @@
+"""
+import json
+
+import EnumsAI
+from freeorion_tools import profile
+
+from _definitions import *
+
+
+@profile
+def load_savegame_string(string):
+    import zlib
+    try:
+        string = zlib.decompress(string)
+    except zlib.error:
+        pass  # probably an uncompressed string
+    return decode(string)
+
+
+def decode(obj):
+    return _FreeOrionAISaveGameDecoder().decode(obj)
+
+
+class _FreeOrionAISaveGameDecoder(json.JSONDecoder):
+
+    def __init__(self, **kwargs):
+        # do not allow control characters
+        super(_FreeOrionAISaveGameDecoder, self).__init__(strict=True, **kwargs)
+
+    def decode(self, s, _w=None):
+        # use the default JSONDecoder to parse the string into a dict
+        # then interpret the dict content according to our encoding
+        retval = super(_FreeOrionAISaveGameDecoder, self).decode(s)
+        return self.__interpret(retval)
+
+    def __interpret_dict(self, obj):
+
+        # if the dict does not contain the class-encoding keys,
+        # then it is a standard dictionary.
+        if not all(key in obj for key in ('__class__', '__module__')):
+            return {self.__interpret(key): self.__interpret(value)
+                    for key, value in obj.iteritems()}
+
+        # pop and verify class and module name, then parse the class content
+        class_name = obj.pop('__class__')
+        module_name = obj.pop('__module__')
+        full_name = '%s.%s' % (module_name, class_name)
+        cls = trusted_classes.get(full_name)
+        if cls is None:
+            raise InvalidSaveGameException("DANGER DANGER - %s not trusted"
+                                           % full_name)
+
+        parsed_content = self.__interpret_dict(obj)
+
+        # create a new instance without calling the actual __new__or __init__
+        # function of the class (so we avoid any side-effects from those)
+        new_instance = object.__new__(cls)
+
+        # Set the content trying to use the __setstate__ method if defined
+        # Otherwise, directly set the dict content.
+        try:
+            setstate = new_instance.__setstate__
+        except AttributeError:
+            if not type(parsed_content) == dict:
+                raise InvalidSaveGameException("Could not set content for %s"
+                                               % new_instance)
+            new_instance.__dict__ = parsed_content
+        else:
+            # only call now to not catch exceptions in the setstate method
+            setstate(parsed_content)
+        return new_instance
+
+    def __interpret(self, x):
+        # primitive types do not have to be interpreted
+        if x is None or isinstance(x, (int, float, bool)):
+            return x
+
+        # special handling for dicts as they could encode our classes
+        if isinstance(x, dict):
+            return self.__interpret_dict(x)
+
+        # for standard containers, interpret each element
+        if isinstance(x, (list, tuple, set)):
```

Comment:
Need some clarification about method argument.  `tuple` and `set` cannot be in `json`.